vpr/src/draw/draw.cpp

```diff
@@ -3997,6 +4066,577 @@ void net_max_fanout(GtkWidget* /*widget*/, gint /*response_id*/, gpointer /*data
+    gtk_grid_attach((GtkGrid*)info_grid, m, 0, 0, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, t, 0, 1, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, i, 2, 0, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, b, 2, 1, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, n, 2, 2, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, r, 0, 2, 1, 1);
+
+    //info grid labels
+    std::string move_num = "move_num: " + std::to_string(current_info_d.move_num);
+    GtkWidget* move_info = gtk_label_new(move_num.c_str());
+    gtk_widget_set_margin_left(move_info, 5);
+    gtk_widget_set_halign(move_info, GTK_ALIGN_START);
+    std::string temp_count = "temp_count: " + std::to_string(current_info_d.temp_count);
+    GtkWidget* temp_info = gtk_label_new(temp_count.c_str());
+    gtk_widget_set_margin_left(temp_info, 5);
+    gtk_widget_set_halign(temp_info, GTK_ALIGN_START);
+    std::string in_blocks_affected = "in_blocks_affected: " + std::to_string(get_current_info_e_ba());
+    GtkWidget* ba_info = gtk_label_new(in_blocks_affected.c_str());
+    gtk_widget_set_halign(ba_info, GTK_ALIGN_START);
+    std::string block_id = "from_block: " + std::to_string(current_info_d.from_block);
+    GtkWidget* block_info = gtk_label_new(block_id.c_str());
+    gtk_widget_set_margin_left(block_info, 5);
+    gtk_widget_set_halign(block_info, GTK_ALIGN_START);
+    std::string router_iter = "router_iter: " + std::to_string(current_info_d.router_iter);
+    GtkWidget* ri_info = gtk_label_new(router_iter.c_str());
+    gtk_widget_set_margin_left(ri_info, 5);
+    gtk_widget_set_halign(ri_info, GTK_ALIGN_START);
+    std::string net_id = "rouet_net_id: " + std::to_string(current_info_d.net_id);
+    GtkWidget* net_info = gtk_label_new(net_id.c_str());
+    gtk_widget_set_margin_left(net_info, 5);
+    gtk_widget_set_halign(net_info, GTK_ALIGN_START);
+    gtk_grid_attach((GtkGrid*)info_grid, move_info, 1, 0, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, temp_info, 1, 1, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, ba_info, 3, 0, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, block_info, 3, 1, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, ri_info, 1, 2, 1, 1);
+    gtk_grid_attach((GtkGrid*)info_grid, net_info, 3, 2, 1, 1);
+    gtk_grid_attach((GtkGrid*)grid, info_grid, 0, 2, 1, 1);
+
+    //button settings
+    GtkWidget* button = gtk_button_new_with_label("OK");
+    gtk_widget_set_margin_bottom(button, 30);
+    gtk_widget_set_halign(button, GTK_ALIGN_CENTER);
+
+    gtk_grid_attach((GtkGrid*)grid, button, 0, 3, 1, 1);
+    g_signal_connect(button, "clicked", G_CALLBACK(ok_close_window), window);
+
+    gtk_container_add(GTK_CONTAINER(window), grid);
+    gtk_widget_show_all(window);
+}
+
+//closess the "invalid entry" window
+void ok_close_window(GtkWidget* /*widget*/, GtkWidget* window) {
+    gtk_window_close((GtkWindow*)window);
+}
+
+//checks if an expression is valid by checking the order of operators
+bool valid_expression(std::string exp) {
```

Comment:
Need to comment what this does and what it is used for.  It's not clear to me what ops being 0 or 1 (magic numbers) means. Might also be better with enumerated constants instead of 1 or 0, but I can't really say since I don't know what this routine is doing.

Reply:
I had assigned 0 to comparing operators and 1 to boolean operators and in a valid expression, they appear in a certain manner. So first I would go through the expression and store the operator codes in a vector, then go through the vector and check for the correct pattern. I changed 0 and 1 to enum types of COMP_OP and BOOL_OP and added more comments. So resolved.